### PR TITLE
Add `zerotensor` design description in code

### PR DESCRIPTION
**Pull request by zeshengzong**
Fix `TODO: add a note explaining the design decisions` by adding design description in https://github.com/pytorch/pytorch/issues/69687 to codebase. Make it easier to get and read by other developers.

### CODE REVIEW

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 1]
```diff
@@ -9,7 +9,77 @@
...
+  * 3. Similar to Conjugate and Negative dispatch keys, ZeroTensor keys also has a
+  *    corresponding boxed fallback for dispatch key resolution. Most functions will
+  *    use this fallback if the ZeroTensor key is set on one or more inputs. A select
+  *    few functions that are registered as fallthrough kernels or have a special
+  *    kernel registered in native_functions.yaml will bypass this fallback (mul,
+  *    matmul, tensor views, etc.).
+  *     - This fallback materializes each ZeroTensor in the following way:
+  *       at::zeros({}, tensor.options()).expand(tensor.sizes()); where tensor is an
+  *       efficient zero tensor (tensor with ZeroTensor dispatch key set)
+  *     - If there’s a mutable ZeroTensor , then we simply error out to avoid the race
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]: Remove this bullet this it is implied by the previous rule.

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 2]
```diff
@@ -9,7 +9,77 @@
...
+  * The goal of ZeroTensors is to replace the undefined tensors used in the forward
+  * and backward mode AD which are currently materialized to a tensor full of zeros
+  * with a new type of zero tensor that use O(1) memory (tensors with a storage
+  * pointing to nullptr). Having undefined tensors is a common occurrence in the
+  * forward mode AD. Currently these undefined tensors are materialized to zero
+  * tensors. Not only the materialization is wasteful, but the cases where
+  * computationally heavy operations like matrix multiplies are called for these
+  * tensors along with some non-zero tensors provide opportunities for perf
+  * improvement. With zero tensors, expensive operations like matmul on zero
+  * tensors will become O(1).
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]: This whole paragraph is more of an introduction to the original issue, for the purpose of this note we do not need this paragraph and can just have a comment at the very bottom of this note saying:
"Today ZeroTensors are primarily used to to represent undefined gradients in forward AD."

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 3]
```diff
@@ -9,7 +9,77 @@
...
+  *
+  * 1. ZeroTensors are regular tensors with TensorOptions (as usual), and a storage
+  *    pointing to nullptr (aliasing semantic is correctly maintained), and a
+  *    ZeroTensor dispatch key set.
+  *
+  * 2. Zero tensors are immutable.
+  *     - As a result, in-place operations on these tensors will be disallowed. This is
+  *       done to prevent data race in the case of multithreading (when two threads try
+  *       to read the same zero tensor and materialize it in-place).
+  *     - Zero tensors shouldn’t be passed as an output tensor in out= operations.
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]: Comment about out= should already be implied.
Just keep the "Zero tensors are immutable" and under that "This is done to prevent data race in the case of multithreading (when two threads try to read the same zero tensor and materialize it in-place)."

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 4]
```diff
@@ -9,7 +9,77 @@
...
+  *
+  * 3. Similar to Conjugate and Negative dispatch keys, ZeroTensor keys also has a
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]: We can summarize this to:
-> ZeroTensor has a boxed fallback that will be dispatched to any ops that don't have special ZeroTensor-handling.  This fallback materializes each ZeroTensor to at::zeros({}, tensor.options()).expand(tensor.sizes()).

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 5]
```diff
@@ -9,7 +9,77 @@
...
+  *     - will never be required for ForwardMode AD.
+  *       - This is because if all the tangents were undefined (efficient zero tensors),
+  *         no computation will be performed (this is ensured via an existing pre-check).
+  *       - If one or more tangent is non-zero (undefined), then even if the Jacobian is
+  *         zero, we never get an efficient zero tensor as an output (since we don’t
+  *         compute efficient zero tensors in Jacobian computation, i.e., Jacobian can
+  *         never be a ZeroTensor). This will not be true anymore if ZeroTensors were
+  *         exposed in the Python API.
+  *
+  * 6. torch.mul(efficient_zero_tensor, other) will return an efficient_zero_tensor.
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]: Let's remove 6/7

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 6]
```diff
@@ -9,7 +9,77 @@
...
+  *    few functions that are registered as fallthrough kernels or have a special
+  *    kernel registered in native_functions.yaml will bypass this fallback (mul,
+  *    matmul, tensor views, etc.).
+  *     - This fallback materializes each ZeroTensor in the following way:
+  *       at::zeros({}, tensor.options()).expand(tensor.sizes()); where tensor is an
+  *       efficient zero tensor (tensor with ZeroTensor dispatch key set)
+  *     - If there’s a mutable ZeroTensor , then we simply error out to avoid the race
+  *       condition (as mentioned above).
+  *
+  * 4. ZeroTensors are handled below autograd. This is necessary because we want the
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]: 4. ZeroTensor are handled above autograd because fallback operations aren't differentiable.

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 7]
```diff
@@ -9,7 +9,77 @@
...
+  *    materialization to happen below the autograd to ensure gradients are populated
+  *    on the correct tensors.
+  *     - Example: Consider add in the case it was using the fallback: zerotensor_a + b.
+  *       zerotensor_a would be materialized to c=torch.zeros_like(zerotensor_a) after
+  *       passing through the fallback. If this happens above the autograd, then the
+  *       gradients would be populated on c instead of zerotensor_a.
+  *
+  * 5. The grad field is always populated with an honest to goodness tensor. This
+  *    materialization of zero tensors will happen in:
+  *     - AcccumulateGrad for Backward Mode AD.
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]:
```suggestion
  *     - AccumulateGrad for Backward Mode AD.
```

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 8]
```diff
@@ -9,7 +9,77 @@
...
+  *     - will never be required for ForwardMode AD.
+  *       - This is because if all the tangents were undefined (efficient zero tensors),
+  *         no computation will be performed (this is ensured via an existing pre-check).
+  *       - If one or more tangent is non-zero (undefined), then even if the Jacobian is
+  *         zero, we never get an efficient zero tensor as an output (since we don’t
+  *         compute efficient zero tensors in Jacobian computation, i.e., Jacobian can
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]: Is this validated anywhere?

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 1 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Changed, thanks!

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 2 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Make this part concise

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 3 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Changed, thanks!

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 4 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Changed, thanks!

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 5 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Changed, thanks!

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 6 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Changed, thanks!

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 7 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Changed, thanks!

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 8 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Removed this part of description.

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 9]
```diff
@@ -9,7 +9,41 @@
 
 namespace at {
 
-  // TODO: add a note explaining the design decisions
+ /*
+  * ZeroTensor, use O(1) memory (tensors with a storage pointing to nullptr),
+  * are primarily used to represent undefined gradients in forward AD.
+  *
+  * ZeroTensors do not propagate nans or other extreme values. This is rooted in
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]: (Thinking about that the div example we were discussing in the other PR) We can actually turn this into something more general->

"ZeroTensor today does not perfectly handle NaNs and Infs as we don't check the actual values and assume that they are non-zero, non-inf, non-NaN etc."

Comment (soulitzer) [reply aten/src/ATen/ZeroTensorFallback.cpp]: We can also put this at the bottom of the note.

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 10]
```diff
@@ -9,7 +9,41 @@
 
 namespace at {
 
-  // TODO: add a note explaining the design decisions
+ /*
+  * ZeroTensor, use O(1) memory (tensors with a storage pointing to nullptr),
+  * are primarily used to represent undefined gradients in forward AD.
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]: Nit: Put the "Today ZeroTensors are primarily used to represent undefined gradients in forward AD" at the end of the note. 

We can then remove the first sentence since it overlaps witih the first bullet of the design.

File: aten/src/ATen/ZeroTensorFallback.cpp [hunk 11]
```diff
@@ -9,7 +9,41 @@
...
+  * ZeroTensors do not propagate nans or other extreme values. This is rooted in
+  * how the kernels are implemented for ZeroTensors. For example, ZT * (non-ZT) = ZT, i.e.,
+  * we don't check whether the non-ZT has nans for example.
+  *
+  * Design:
+  *
+  * 1. ZeroTensors are regular tensors with TensorOptions, a storage
+  *    pointing to nullptr and a ZeroTensor dispatch key set.
+  *
+  * 2. ZeroTensors are immutable.
```
Comment (soulitzer) [inline aten/src/ATen/ZeroTensorFallback.cpp]: nit: For any bullets that only have one sub bullet, let's collapse that into just the main bullet. 

e.g. 

"2. ZeroTensors are immutable. This is done to prevent ..."

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 9 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Changed, combined with first paragraph and put in bottom of note, thanks!

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 10 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Changed, thanks!

File: aten/src/ATen/ZeroTensorFallback.cpp (hunk 11 above)
Comment (zeshengzong) [reply aten/src/ATen/ZeroTensorFallback.cpp]: Changed, thanks!

**Review by soulitzer — APPROVED** [review]
Thanks!
